internal/feature.go

```diff
@@ -85,26 +86,43 @@ func NewScriptBasedSU(scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig) (*Scr
 	}
 
 	// Get the local edge configuration.
-	edge, mc, err := retrieveEdgeConfiguration(scriptSUPConfig.Broker, scriptSUPConfig.Username, scriptSUPConfig.Password)
+	edge, err := NewEdgeConnector(scriptSUPConfig, feature)
 	if err != nil {
 		return nil, err
 	}
-	feature.mqttClient = mc
-	return feature, feature.init(scriptSUPConfig, edge)
+	return edge, nil
 }
 
 // Connect the client to the configured Ditto endpoint.
 // If any error occurs during the connection's initiation - it's returned here.
-func (f *ScriptBasedSoftwareUpdatable) Connect() error {
+func (f *ScriptBasedSoftwareUpdatable) Connect(client MQTT.Client, scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig, edgeCfg *edgeConfiguration) error {
+	logger.Infof("Connecting to ditto endpoint with configuration - %v", edgeCfg)
+	f.mqttClient = client
+	done = make(chan struct{})
+	f.queue = make(chan operationFunc, 10)
+	f.lock.Lock()
+	connected = true
```

Comment:
It seems like a race condition. the flag will indicate that the feature is connected but it's not. Connect and disconnect has to be in synch in the context of their initialization.